### ISSUE TITLE
Support installation with Pydantic v2

### ISSUE THREAD

**Issue by jawnsy** [PrefectHQ/prefect#10145]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar request and didn't find it.
- [X] I searched the Prefect documentation for this feature.

### Prefect Version

2.x

### Describe the current behavior

We are currently using Pydantic v1 (and [we have it pinned](https://github.com/PrefectHQ/prefect/pull/10144) to prevent build failures).

### Describe the proposed behavior

Ideally, we would like to support installation with both v1 and v2, until v1 falls out of support. This would require us to understand the [forward compatibility features available in v1](https://github.com/pydantic/pydantic/issues/5165) and adopt those.

**Comment by leonardodiegues** [comment]
Do you guys need help with the V2 installation? I'm willing to make a PR addressing this issue if you like

**Comment by jawnsy** [comment]
@leonardodiegues Thanks for the offer! We'd be happy to review a pull request that adds support for installation with Pydantic v1 as well as Pydantic v2 alongside Prefect. I'm not too familiar with the required changes yet, can you describe an approach here so we can discuss prior to your implementation?

**Comment by leonardodiegues** [comment]
You're right, we should discuss things before doing things.

I think the upgrade doesn't take much effort. They basically migrated things from one place to another, but all functionalities are preserved with slightly different names and implementations.

They published a detailed [migration guide](https://docs.pydantic.dev/2.0/migration/) that should guide us.

I'm currently checking which changes we will have to deal amongst all breaking features and will post here asap.

**Comment by leonardodiegues** [comment]
FastAPI is currently the only Prefect dependency that conflicts with Pydantic V2. They're discussing it [here](https://github.com/tiangolo/fastapi/issues/9710)

**Comment by zanieb** [comment]
I investigated upgrading to Pydantic v2 previously and gave up when I realized FastAPI did not have support yet.

We also do some complicated things with Pydantic that were hard to translate to v2, such as our `FieldFrom` / `copy_model_fields` utilities and model deserialization dispatch.

**Comment by leonardodiegues** [comment]
> I investigated upgrading to Pydantic v2 previously and gave up when I realized FastAPI did not have support yet (done in [tiangolo/fastapi#9500](https://github.com/tiangolo/fastapi/pull/9500) & more now). Unfortunately I can't find the branch.
> 
> We also do some complicated things with Pydantic that were hard to translate to v2, such as our `FieldFrom` / `copy_model_fields` utilities and model deserialization dispatch.

The changes make it hard to check on field validators/serializers from the field perspective since both can only be accessed through the model class if implemented in the form of decorators. One solution would be capture/set these guys through `BaseModel.__pyarrow_decorators__`.

Other way would be to re-type class attributes in `typing.Annotated` style, like they mention in the migration guide. It's easier to refer the validation methods because they can be directly accessed through `FieldInfo.annotation`, but It would take too long to migrate everything from what we have now to this solution.

There could be some other alternative I'm not seeing here.

**Comment by Kludex** [comment]
If help is needed, feel free to reach me out. 🙏

**Comment by jcorrado76** [comment]
Just to add another piece of information: When I was trying to install Prefect using Poetry, it installed, but import failed. I was getting an `ImportError: Cannot import SecretField from pydantic`. I had to pin `pydantic="^1.0"` or `pydantic<2`. I see this in the `requirements.txt` as well.

**Comment by leonardodiegues** [comment]
This should help us: https://github.com/pydantic/bump-pydantic

**Comment by sckott** [comment]
I'm running into this error `ImportError: Cannot import SecretField from pydantic` when running `prefect` on GitHub Actions. Any solution for now? Does this not happen with some older or newer versions of `prefect`?

**Comment by jcorrado76** [comment]
Hi @sckott - this happened to me as well. Mine was solved by pinning pydantic <2. 

Cheers!

**Comment by discdiver** [comment]
[sckott](https://github.com/sckott) it happens with some older versions of Prefect, but you shouldn't see it with the most recent versions (e.g. 2.10.20). If you do see it with your versions of Prefect you can pin pydantic as @jcorrado76 suggests.

**Comment by sckott** [comment]
@discdiver thanks. yeah i upgraded and its all good now

**Comment by ybressler** [comment]
You can maintain compatibility by changing the import path. Something like:
```py3
try:  # if you are in V2, import things form ehre
    import pydantic.v1 as pydantic

except:  # otherwise, use it normally
    import pydantic
```

**Comment by frctnlss** [comment]
@discdiver would it be possible to have an announcement in the community channel when discoveries like this are found? I am not sure how other companies are managing their Prefect versions, but we don't upgrade unless there is a fix or feature we need. We also have our runtime decoupled from our deployment, so the only time we would find this is when we make a change that requires a new deployment be executed. Our company, Datateer, leverages prefect cloud instead of running the API server ourselves, but the deployments and agents are still something we need to self-manage. I do have automated notifications set up when there are new releases of prefect and review them, but this pydantic issue was not expressed as a breaking change for older versions. We were on 2.10.8 prior to today.

**Comment by FlowMatric** [comment]
@cicdw , any idea when this might make it's way into a release?

**Comment by Kludex** [comment]
If there's something we can do to help. Let us know. 🙏

**Comment by dgasmith** [comment]
Weighing in here- it would be fantastic to upgrade and happy to help.

**Comment by cicdw** [comment]
Hey everyone - heads up we're prioritizing this now and are beginning the prep work with a goal of releasing by end of October.  We think we can get it done well before then, but I want to add some buffer just because of the scope of the changes.  We'll keep this issue updated in the meantime!

**Comment by urimandujano** [comment]
Hi everyone, we released `prefect v2.13.5` with support for pydantic v2 (via forced installation) and then removed our pin on pydantic 1 in `prefect v2.13.7`. Today we released `prefect v2.14.3` which fixes some internal incompatibilities with  pydantic v2 schemas. 

At this point, we're compatible with environments that have pydantic v1 or v2 installed. Please give it a shot and open new issues if you have any problems. Thanks for the support!
